### PR TITLE
 Feature/udp proxy

### PR DESCRIPTION
**Pull request by dengqian**
### Issues associated with this PR

Support UDP proxy

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/configmanager/parser.go [hunk 1]
```diff
@@ -185,23 +210,51 @@ func ParseListenerConfig(lc *v2.Listener, inheritListeners []net.Listener) *v2.L
 			log.StartLogger.Fatalf("[config] [parse listener] inheritListener not valid: %s", tl.Addr().String())
 		}
 
-		if addr.Port != ilAddr.Port {
+		if GetAddrPort(addr) != ilAddr.Port {
```
Comment (taoyuanyuan) [inline pkg/configmanager/parser.go:213]: It should not be compared with tcp，if the addr is udp

File: pkg/configmanager/parser.go [hunk 2]
```diff
@@ -185,23 +210,51 @@ func ParseListenerConfig(lc *v2.Listener, inheritListeners []net.Listener) *v2.L
...
+		if il == nil {
+			continue
+		}
+		tl := il.(*net.UDPConn)
+		ilAddr, err := net.ResolveUDPAddr("udp", tl.LocalAddr().String())
+		if err != nil {
+			log.StartLogger.Fatalf("[config] [parse listener] inheritListener not valid: %s", tl.LocalAddr().String())
+		}
+
+		if GetAddrPort(addr) != ilAddr.Port {
```
Comment (taoyuanyuan) [inline pkg/configmanager/parser.go:239]: ditto

File: pkg/configmanager/parser.go [hunk 3]
```diff
@@ -185,23 +210,51 @@ func ParseListenerConfig(lc *v2.Listener, inheritListeners []net.Listener) *v2.L
...
+			old_pc = il
+			inheritPacketConn[i] = nil
+			break
+		}
+	}
+
 	lc.Addr = addr
 	lc.PerConnBufferLimitBytes = 1 << 15
 	lc.InheritListener = old
+	lc.InheritPacketConn = &old_pc
```
Comment (taoyuanyuan) [inline pkg/configmanager/parser.go:257]: &old_pc not equal nil

File: pkg/configmanager/parser.go [hunk 4]
```diff
@@ -163,12 +163,37 @@ func ParseLogLevel(level string) log.Level {
...
+		return addr.(*net.TCPAddr).IP
+	}
+}
+
+func GetAddrPort(addr net.Addr) int {
+	switch addr.Network() {
+	case "udp":
+		return addr.(*net.UDPAddr).Port
+	default:
+		return addr.(*net.TCPAddr).Port
```
Comment (taoyuanyuan) [inline pkg/configmanager/parser.go:180]: It will panic if addr is unixsocket

File: pkg/network/listener.go [hunk 5]
```diff
@@ -69,13 +71,23 @@ func NewListener(lc *v2.Listener) types.Listener {
...
+		network:                 lc.Network,
 		config:                  lc,
 	}
 
 	if lc.InheritListener != nil {
 		//inherit old process's listener
 		l.rawl = lc.InheritListener
 	}
+
+	if lc.InheritPacketConn != nil {
```
Comment (taoyuanyuan) [inline pkg/network/listener.go:83]: InheritPacketConn is not nil always

File: pkg/filter/network/streamproxy/streamproxy.go [hunk 6]
```diff
@@ -351,15 +384,34 @@ func NewProxyConfig(config *v2.TCPProxy) ProxyConfig {
...
 
+func (pc *proxyConfig) GetIdleTimeout(network string) time.Duration {
+	if pc.idleTimeout != nil {
+		return *pc.idleTimeout
+	} else if strings.Contains(network, "udp") {
+		return types.DefaultUDPIdleTimeout
+	} else {
+		return types.DefaultIdleTimeout
+	}
+}
```
Comment (taoyuanyuan) [inline pkg/filter/network/streamproxy/streamproxy.go:395]: code style	
```
        if pc.idleTimeout != nil {
		return *pc.idleTimeout
	}
	if strings.Contains(network, "udp") {
		return types.DefaultUDPIdleTimeout
	}
	return types.DefaultIdleTimeout
```

File: pkg/server/handler.go [hunk 7]
```diff
@@ -810,53 +835,61 @@ func GetInheritListeners() ([]net.Listener, net.Conn, error) {
...
 	}
 	gotFds, err := unix.ParseUnixRights(&scms[0])
 	if err != nil {
 		log.StartLogger.Errorf("[server] unix.ParseUnixRights: %v", err)
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	listeners := make([]net.Listener, len(gotFds))
+	var packetConn []net.PacketConn
```
Comment (taoyuanyuan) [inline pkg/server/handler.go:864]: listeners keep style consistency about packetConn

File: pkg/filter/network/streamproxy/streamproxy.go [hunk 8]
```diff
@@ -250,10 +271,14 @@ type IpRangeList struct {
 }
 
 func (ipList *IpRangeList) Contains(address net.Addr) bool {
-	tcpAddr, ok := address.(*net.TCPAddr)
-	log.DefaultLogger.Tracef("IpRangeList check ip = %v,address = %v", tcpAddr, address)
-	if ok {
-		ip := tcpAddr.IP
+	var ip net.IP
+	if tcpAddr, ok := address.(*net.TCPAddr); ok {
```
Comment (taoyuanyuan) [inline pkg/filter/network/streamproxy/streamproxy.go:275]: code style, switch

File: pkg/filter/network/streamproxy/streamproxy.go [hunk 9]
```diff
@@ -269,9 +294,17 @@ type PortRangeList struct {
...
 func (pr *PortRangeList) Contains(address net.Addr) bool {
-	tcpAddr, ok := address.(*net.TCPAddr)
-	if ok {
-		port := tcpAddr.Port
+	var port int
+	port = 0
+	if tcpAddr, ok := address.(*net.TCPAddr); ok {
+		port = tcpAddr.Port
+	} else {
+		if udpAddr, ok1 := address.(*net.UDPAddr); ok1 {
```
Comment (taoyuanyuan) [inline pkg/filter/network/streamproxy/streamproxy.go:302]: switch

File: pkg/filter/network/streamproxy/streamproxy.go [hunk 10]
```diff
@@ -351,15 +384,34 @@ func NewProxyConfig(config *v2.TCPProxy) ProxyConfig {
...
+	} else if strings.Contains(network, "udp") {
+		return types.DefaultUDPIdleTimeout
+	} else {
+		return types.DefaultIdleTimeout
+	}
+}
+
+func (pc *proxyConfig) GetReadTimeout(network string) time.Duration {
+	switch network {
+	case "udp", "udp4", "udp6":
```
Comment (taoyuanyuan) [inline pkg/filter/network/streamproxy/streamproxy.go:399]: delete  udp4, udp6

File: pkg/filter/network/streamproxy/streamproxy.go [hunk 11]
```diff
@@ -101,12 +103,21 @@ func (p *proxy) InitializeReadFilterCallbacks(cb api.ReadFilterCallbacks) {
 	// TODO: set downstream connection stats
 }
 
+func (p *proxy) getUpstreamConnection(ctx types.LoadBalancerContext, snapshot types.ClusterSnapshot) types.CreateConnectionData {
+	switch p.network {
+	case "udp", "udp4", "udp6":
```
Comment (taoyuanyuan) [inline pkg/filter/network/streamproxy/streamproxy.go:108]: delete udp4, udp6

File: pkg/network/connection.go [hunk 12]
```diff
@@ -232,11 +248,23 @@ func (c *connection) attachEventLoop(lctx context.Context) {
 }
 
 func (c *connection) checkUseWriteLoop() bool {
-	tcpAddr, ok := c.remoteAddr.(*net.TCPAddr)
-	if !ok {
-		return false
+	var ip net.IP
+	switch c.network {
+	case "udp", "udp4", "udp6":
```
Comment (taoyuanyuan) [inline pkg/network/connection.go:253]: delete udp4,udp6

File: pkg/network/connection.go [hunk 13]
```diff
@@ -642,6 +702,9 @@ func (c *connection) startWriteLoop() {
 				c.Close(api.NoFlush, api.LocalClose)
 			}
 
+			if strings.Contains(err.Error(), "connection refused"){
```
Comment (taoyuanyuan) [inline pkg/network/connection.go:705]: readloop will close connection

File: pkg/network/connection.go [hunk 14]
```diff
@@ -683,7 +746,26 @@ func (c *connection) doWriteIo() (bytesSent int64, err error) {
 		bytesSent, err = tlsConn.WriteTo(&buffers)
 	} else {
 		//todo: writev(runtime) has memroy leak.
-		bytesSent, err = buffers.WriteTo(c.rawConnection)
+		if strings.Contains(c.network, "tcp") {
```
Comment (taoyuanyuan) [inline pkg/network/connection.go:749]: switch

File: pkg/network/udp_recv.go [hunk 15]
```diff
@@ -0,0 +1,88 @@
+/*
```
Comment (taoyuanyuan) [inline pkg/network/udp_recv.go:1]: the filename is nginx style, maybe change it.

File: pkg/network/udp_recv.go [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+
+	"mosn.io/api"
+	"mosn.io/mosn/pkg/log"
+	"mosn.io/pkg/buffer"
+)
+
+const UdpPacketMaxSize = 64 * 1024
+
+var (
+	bufPool  =  buffer.IoBufferPool{}
```
Comment (taoyuanyuan) [inline pkg/network/udp_recv.go:34]: unused

File: pkg/network/udp_recv.go [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+func GetProxyMapKey(raddr, laddr string) string {
+	var builder strings.Builder
+	builder.WriteString(raddr)
+	builder.WriteString(":")
+	builder.WriteString(laddr)
+	return builder.String()
+}
+
+func SetUdpProxyMap(key string, conn api.Connection) {
```
Comment (taoyuanyuan) [inline pkg/network/udp_recv.go:46]: SetUDPProxyMap

File: pkg/network/udp_recv.go [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+	ProxyMap.Store(key, conn)
+}
+
+func DelUdpProxyMap(key string) {
```
Comment (taoyuanyuan) [inline pkg/network/udp_recv.go:50]: DelUDPProxyMap

File: pkg/network/connection.go [hunk 19]
```diff
@@ -642,6 +702,9 @@ func (c *connection) startWriteLoop() {
 				c.Close(api.NoFlush, api.LocalClose)
 			}
 
+			if c.network == "udp" && strings.Contains(err.Error(), "connection refused") {
```
Comment (taoyuanyuan) [inline pkg/network/connection.go:705]: other write errs not close connection, beacause readbuffer may have unread data, wait for readloop close connection

File: pkg/network/udpreceiver.go [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+	conn := l.packetConn.(*net.UDPConn)
+	buf := buffer.GetIoBuffer(UdpPacketMaxSize)
+	defer buffer.PutIoBuffer(buf)
+	for {
+		buf.Reset()
+		n, rAddr, err := conn.ReadFromUDP(buf.Bytes()[:buf.Cap()])
+		buf.Grow(n)
+
+		if err != nil {
+			if nerr, ok := err.(net.Error); ok && (nerr.Timeout() || nerr.Timeout()) {
```
Comment (taoyuanyuan) [inline pkg/network/udpreceiver.go:63]: dup nerr.Timeout()，and  why break when timeout？

File: go.mod1 [hunk 21]
```diff
@@ -0,0 +1,10 @@
+module mosn.io/mosn
```
Comment (taoyuanyuan) [inline go.mod1:1]: tmp file？

File: pkg/config/v2/server.go [hunk 22]
```diff
@@ -113,7 +115,7 @@ func (fc *FilterChain) UnmarshalJSON(b []byte) error {
 	if len(fc.TLSConfigs) > 0 {
 		fc.TLSContexts = make([]TLSConfig, len(fc.TLSConfigs))
 		copy(fc.TLSContexts, fc.TLSConfigs)
-	} else { // no tls_context_set, use tls_context
+	} else {                     // no tls_context_set, use tls_context
```
Comment (taoyuanyuan) [inline pkg/config/v2/server.go:118]: replace with space

File: pkg/configmanager/parser.go [hunk 23]
```diff
@@ -163,45 +163,109 @@ func ParseLogLevel(level string) log.Level {
...
+			tl := il.(*net.TCPListener)
+			ilAddr, err := net.ResolveTCPAddr("tcp", tl.Addr().String())
+			if err != nil {
+				log.StartLogger.Fatalf("[config] [parse listener] inheritListener not valid: %s", tl.Addr().String())
+			}
 
-		if addr.Port != ilAddr.Port {
-			continue
-		}
+			if GetAddrPort(addr) != ilAddr.Port {
```
Comment (taoyuanyuan) [inline pkg/configmanager/parser.go:247]: replace with space

File: go.mod [hunk 24]
```diff
@@ -49,8 +49,8 @@ require (
 	gopkg.in/yaml.v2 v2.2.8 // indirect
 	istio.io/api v0.0.0-20200227213531-891bf31f3c32
 	k8s.io/klog v1.0.0
-	mosn.io/api v0.0.0-20200728020814-8cdc3e8890f9
-	mosn.io/pkg v0.0.0-20200723115157-3842674e37c7
+  mosn.io/api v0.0.0-20200729124336-c71e8f2074cb
+  mosn.io/pkg v0.0.0-20200729115159-2bd74f20be0f
```
Comment (taoyuanyuan) [inline go.mod:53]: fmt

**Comment by taoyuanyuan** [comment]
please add an example such as `examples/cn_readme/tcpproxy-sample`

File: examples/cn_readme/udpproxy-sample/README.md [hunk 25]
```diff
@@ -0,0 +1,60 @@
+## 使用 MOSN 作为TCP 代理
```
Comment (taoyuanyuan) [inline examples/cn_readme/udpproxy-sample/README.md:1]: TCP ->UDP

File: examples/cn_readme/udpproxy-sample/README.md [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+
+## 简介
+
++ 该样例工程演示了如何配置使得MOSN作TCP Proxy代理
```
Comment (taoyuanyuan) [inline examples/cn_readme/udpproxy-sample/README.md:5]: UDP

File: examples/cn_readme/udpproxy-sample/README.md [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
++ MOSN收到一个TCP请求，会根据请求得源地址、目的地址(不配置则为任意地址)转发到对应的cluster
```
Comment (taoyuanyuan) [inline examples/cn_readme/udpproxy-sample/README.md:6]: UDP

File: examples/en_readme/udpproxy-sample/README.md [hunk 28]
```diff
@@ -0,0 +1,60 @@
+## Use MOSN as TCP Proxy
```
Comment (taoyuanyuan) [inline examples/en_readme/udpproxy-sample/README.md:1]: UDP

File: examples/en_readme/udpproxy-sample/README.md [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+
+## Introduction
+
++ This sample project demonstrates how to configure MOSN as a TCP Proxy.
```
Comment (taoyuanyuan) [inline examples/en_readme/udpproxy-sample/README.md:5]: UDP

File: pkg/network/listener.go [hunk 30]
```diff
@@ -133,40 +145,69 @@ func (l *listener) Start(lctx context.Context, restart bool) {
...
-					// stop accepting loop and log the event
-					if !(ope.Timeout() && ope.Temporary()) {
-						// accept error raised by sockets closing
-						if ope.Op == "accept" {
-							log.DefaultLogger.Infof("[network] [listener start] [accept] listener %s %s closed", l.name, l.Addr())
-						} else {
-							log.DefaultLogger.Alertf("listener.accept", "[network] [listener start] [accept] listener %s occurs non-recoverable error, stop listening and accepting:%s", l.name, err.Error())
-						}
-						return
+func (l *listener) AcceptEventLoop(lctx context.Context) {
```
Comment (taoyuanyuan) [inline pkg/network/listener.go:169]: AcceptEventLoop -> acceptEventLoop

File: pkg/network/listener.go [hunk 31]
```diff
@@ -133,40 +145,69 @@ func (l *listener) Start(lctx context.Context, restart bool) {
...
+					return
 				}
+			} else {
+				log.DefaultLogger.Errorf("[network] [listener start] [accept] listener %s occurs unknown error while accepting:%s", l.name, err.Error())
 			}
 		}
 	}
 }
 
+func (l *listener) ReadMsgEventLoop(lctx context.Context) {
```
Comment (taoyuanyuan) [inline pkg/network/listener.go:194]: ReadMsgEventLoop -> readMsgEventLoop

File: pkg/network/listener.go [hunk 32]
```diff
@@ -133,40 +145,69 @@ func (l *listener) Start(lctx context.Context, restart bool) {
...
+	utils.GoWithRecover(func() {
+		ReadMsgLoop(lctx, l)
```
Comment (taoyuanyuan) [inline pkg/network/listener.go:196]: ReadMsgLoop -> readMsgLoop
